pkg/controller/importconfig/importconfig_controller.go

```diff
@@ -96,56 +56,106 @@ func (r *ReconcileImportConfig) Reconcile(ctx context.Context, request reconcile
+	if err != nil {
+		return reconcile.Result{}, err
 	}
-
 	if _, err := helpers.ApplyResources(
 		r.clientHolder, r.recorder, r.scheme, managedCluster, objects...); err != nil {
 		return reconcile.Result{}, err
 	}
 
-	// make sure the managed cluster import secret is updated
-	importSecret, err := worker.generateImportSecret(ctx, managedCluster)
+	// get the previous bootstrap kubeconfig and expiration
+	bootstrapKubeconfigData, expiration, err := getBootstrapKubeConfigDataFromImportSecret(ctx, r.clientHolder, managedCluster.Name)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
+	// if bootstrapKubeconfig not exist or expired, create a new one
+	if bootstrapKubeconfigData == nil {
+		bootstrapKubeconfigData, expiration, err = bootstrap.CreateBootstrapKubeConfig(ctx, r.clientHolder,
+			bootstrap.GetBootstrapSAName(managedCluster.Name), managedCluster.Name, 8640*3600)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	var yamlcontent, crdsV1YAML, crdsV1beta1YAML []byte
+	var secretAnnotations map[string]string
+	switch mode {
+	case constants.KlusterletDeployModeDefault:
+		yamlcontent, err = bootstrap.NewKlusterletManifestsConfig(
+			operatorv1.InstallModeDefault,
+			managedCluster.Name,
+			klusterletNamespace(managedCluster.GetAnnotations()),
+			bootstrapKubeconfigData).
+			WithManagedClusterAnnotations(managedCluster.GetAnnotations()).
+			Generate(ctx, r.clientHolder)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		crdsV1beta1YAML, err = bootstrap.GenerateKlusterletCRDsV1Beta1()
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		crdsV1YAML, err = bootstrap.GenerateKlusterletCRDsV1()
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	case constants.KlusterletDeployModeHosted:
+		yamlcontent, err = bootstrap.NewKlusterletManifestsConfig(
+			operatorv1.InstallModeHosted,
+			managedCluster.Name,
+			klusterletNamespace(managedCluster.GetAnnotations()),
+			bootstrapKubeconfigData,
+		).
+			WithManagedClusterAnnotations(managedCluster.GetAnnotations()).
+			WithImagePullSecretGenerate(false).Generate(ctx, r.clientHolder)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		secretAnnotations = map[string]string{
+			constants.KlusterletDeployModeAnnotation: constants.KlusterletDeployModeHosted,
+		}
+	default:
+		return reconcile.Result{}, fmt.Errorf("klusterlet deploy mode %s not supportted", mode)
+	}
+
+	// generate import secret
+	importSecret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", managedCluster.Name, constants.ImportSecretNameSuffix),
+			Namespace: managedCluster.Name,
+			Labels: map[string]string{
+				constants.ClusterImportSecretLabel: "",
+			},
+			Annotations: secretAnnotations,
+		},
+		Data: map[string][]byte{
+			constants.ImportSecretImportYamlKey:      yamlcontent,
+			constants.ImportSecretCRDSYamlKey:        crdsV1YAML,
```

Comment:
@skeeey do you remember why we need two keys to store the same crdsv1 yaml? is it only for backward compatible, can we deduplicate it?

Reply:
yes, it's only for backward compatible, we need support opc311